std/typecons.d

```diff
@@ -3869,15 +3869,119 @@ unittest
+        // list of function symbols
+        alias SourceMembers = GetOverloadedMethods!Source;
+
+        // Check whether all of SourceMembers satisfy covariance target in
+        // TargetMembers
+        template hasRequireMethods(size_t i = 0)
+        {
+            static if (i >= TargetMembers.length)
+                enum hasRequireMethods = true;
+            else
+            {
+                enum hasRequireMethods =
+                    findCovariantFunction!(TargetMembers[i], Source, SourceMembers) != -1 &&
+                    hasRequireMethods!(i + 1);
+            }
+        }
+    }
+}
+/// ditto
+private template implementsInterface(Source, Targets...)
+if (Targets.length >= 1 && !allSatisfy!(isMutable, Targets))
+{
+    import std.typetuple : staticMap;
+
+    alias implementsInterface = .implementsInterface!(Source, staticMap!(Unqual, Targets));
+}
+
+unittest {
+    interface Foo {
+        void foo();
+    }
+    interface Bar {
+        void bar();
+    }
+    interface FooBar : Foo, Bar {
+        void foobar();
+    }
+
+    struct A {
+        void foo() {}
+    }
+    struct B {
+        void bar() {}
+        void foobar() {}
+    }
+    class C {
+        void foo() {}
+        void bar() {}
+    }
+    struct D {
+        void foo() {}
+        void bar() {}
+        void foobar() {}
+    }
+    // Implements interface
+    static assert(implementsInterface!(A, Foo));
+    static assert(implementsInterface!(A, const(Foo)));
+    static assert(implementsInterface!(A, immutable(Foo)));
+    // Doesn't implement interface
+    static assert(!implementsInterface!(B, Foo));
+    static assert(implementsInterface!(B, Bar));
+    // Implements both interfaces
+    static assert(implementsInterface!(C, Foo));
+    static assert(implementsInterface!(C, Bar));
+    static assert(implementsInterface!(C, Foo, Bar));
+    static assert(implementsInterface!(C, Foo, const(Bar)));
+    static assert(!implementsInterface!(A, Foo, Bar));
+    static assert(!implementsInterface!(A, Foo, immutable(Bar)));
+    // Implements inherited
+    static assert(implementsInterface!(D, FooBar));
+    static assert(!implementsInterface!(B, FooBar));
+}
+
+private template isInterface(ConceptType) {
+    enum isInterface = is(ConceptType == interface);
+}
+
 /**
  * Supports structural based typesafe conversion.
  *
- * If $(D Source) has structural conformance with the $(D interface) $(D Targets),
- * wrap creates internal wrapper class which inherits $(D Targets) and
- * wrap $(D src) object, then return it.
+ * If $(D Source) has structural conformance with the $(D interface)
+ * $(D Targets), wrap creates internal wrapper class which inherits
+ * $(D Targets) and wrap $(D src) object, then return it.
  */
 template wrap(Targets...)
-if (Targets.length >= 1 && allSatisfy!(isMutable, Targets))
+if (Targets.length >= 1 && allSatisfy!(isMutable, Targets)
+    && allSatisfy!(isInterface, Targets))
```

Comment:
Since this will be part of the public interface, how about this. Make the template constraint `if (Targets.length >= 1 && allSatisfy!(isInterface, Targets))` and dispatch inside on `allSatisfy!(isMutable, Targets)` which is an implementation detail.